\documentclass{bioinfo}
\copyrightyear{2019} \pubyear{2019}

\access{Advance Access Publication Date: Day Month 2019}
\appnotes{Application Note}

\begin{document}
\firstpage{1}

\subtitle{Genetic and population analysis}

\title[MPCC]{A Performant Matrix of Pearson's Correlation Coefficient (MPCC) Calculations with Missing Data Support}
\author[Arends \textit{et~al}.]{
Danny Arends\,$^{\text{\sfb 1, $\dagger$}}$, 
Mitch Horton\,$^{\text{\sfb 2, $\dagger$}}$ 
Chad Burdyshaw\,$^{\text{\sfb 2, $\dagger$}}$ 
Christian Fischer\,$^{\text{\sfb 3}}$ 
Pjotr Prins\,$^{\text{\sfb 3}}$ 
Rob W. Williams\,$^{\text{\sfb 3, *}}$ 
, and Glen Brook\,$^{\text{\sfb 2, *}}$}
\address{$^{\text{\sf 1}}$Z{\"u}chtungsbiologie und molekulare 
Genetik, Albrecht Daniel Thaer-Institut, Berlin, 10115, Germany \\
$^{\text{\sf 2}}$The Joint Institute for Computational Sciences, 
University of Tennessee, Oak 
Ridge, TN 37830, USA\\
$^{\text{\sf 3}}$Genetics, Genomics and Informatics, University 
of Tennessee Health Science Center, Memphis, TN 38163, USA.}

\corresp{$^\dagger$Contributed equally and should be considered 
joined first authors, $^\ast$To whom correspondence should be 
addressed.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{\textbf{Motivation:} 
The work presented is motivated by GeneNetwork.org, which performs 
a matrix of Pearson$'$s Correlation Coefficient (PCC) calculations 
in the presence of missing data to find relationships between and 
among genotypes and phenotypes in mouse strains. The calculations 
are a bottleneck for moderate to large problem sizes. Calculating 
PCC is pervasive across bioinformatics, data analysis, 
phylogenetics, statistics, stochastics, and xanthropology. Our 
approach can be used anywhere a matrix of PCC calculations is 
computed.\\
\textbf{Results:} Our solution is a reformulation of the algorithm 
such that it is based on the matrix-matrix product algorithm and 
achieves 4.3 TFlop/s in single precision (77\% of the theoretical 
peak) on a single Intel Xeon Gold 6148 CPU $@$ 2.4 GHz (Skylake) 
compute node. This translates to as much as a 90x speedup over the 
previous approach and is independent of the percentage of missing 
data.\\
\textbf{Availability:} Code is available under an GPL-v3 
licence for C++ and The R Project for Statistical 
Computing at \href{https://github.com/UTennessee-JICS/MPCC}{https://github.com/UTennessee-JICS/MPCC}\\
\textbf{Contact:} \href{rwilliams@uthsc.edu}{rwilliams@uthsc.edu} or 
\href{glenn-brook@tennessee.edu}{glenn-brook@tennessee.edu}\\
\textbf{Supplementary information:} Supplementary data are 
available at \textit{Bioinformatics} online.}

\maketitle

\section{Introduction}
Biology has driven innovation in statistics from the early days. 
Pearson, Fischer, Galton, and many other now famous statisticians 
were all working on biological data. The mathematical formula 
for Pearson correlation were derived by Auguste Bravais in 1844. 
However, as Stigler's Law \citep{Stigler1980} dictates the name 
of the method is crediting Karl Pearson, who was building on 
ideas published by Francis Galton in the 1880s. 

Pearson correlation is one of the most used correlation algorithms in science. 
It is used ubiquitously in all field of science ranging from agriculture to 
zoology. Examples of large scale correlation computation can be found in many 
areas of biology and bioinformatics. Genotype correlation are commonly used to 
construct haplotypes, build genetic maps, and order markers within the genome. 
Furthermore, PCCs have been used in co-expression analysis \citep{Tesson:2010}, 
(genome wide) association analysis, and novel algortihms to reconstruct genetic 
networks such as DiffCor \citep{Fukushima:2013}, weighted correlation network 
analysis (WGCNA) \citep{Horvath:2008} and correlated trait locus (CTL) mapping 
\citep{Arends2016a}.

The BxD family currently consists of 150 inbred mice, and the BxD phenome consists 
of 7000 hand curated and well-integrated classical phenotypes, and over 100 'omics' 
datasets. Almost all BXD expression and phenotype data is freely available on 
\href{https://genenetwork.org/}{GeneNetwork.org} \citep{Sloan2016}, which provides 
powerful statistical tools for data analysis, including Pearson correlation.

PCCs computation in the R language for statistical computing \citep{R:2005} is 
provided by the $cor()$ function. The computation is implemented in C / C++ and 
is relatively performant when no missing data is present. However, 
in the presence of missing data 5 different options on how-to deal 
with missing data are provided. The main choices are removes all rows that 
contain missing data, essentially creating a dataset without missing data. This is 
unwanted in bioinformatics because missing data is often omni-present. 
The second main choice is to handle missing data on a case by case basis. However, 
due to the implementation this causes branching on missing data, leading to CPU 
cache inefficiencies, significanly increasing the runtime of the computation.
%\enlargethispage{12pt}

\section{Approach}
We optimized the PCC algorithm and created a 'drop in' replacement for the $cor()$ 
function provided by R. Optimizations performed are itemized below, and further 
explained in the methods section.

\begin{itemize}
\item Missing data masking
\item Memory access optimizations
\item Threading
\item Vectorization
\item Matrix implementation
\end{itemize}

\begin{methods}
\section{Methods}

We started with the naive version (a slightly adjusted version compared to the R 
implementation) of the algorithm, a series which consists of vector-vector 
operations inside a double nested loop. The problem was reformulated into a 
series of matrix-matrix operations.

\subsection{Missing data masking}
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a 
nibh pulvinar, ultricies sapien eget, consectetur massa. In augue 
ante, iaculis et justo et, imperdiet consectetur nisl. In magna 
nisl, aliquam vel faucibus ac, dapibus ut nisl. 

\subsection{Memory access optimizations}
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a 
nibh pulvinar, ultricies sapien eget, consectetur massa. In augue 
ante, iaculis et justo et, imperdiet consectetur nisl. In magna 
nisl, aliquam vel faucibus ac, dapibus ut nisl. 

\subsection{Threading}
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a 
nibh pulvinar, ultricies sapien eget, consectetur massa. In augue 
ante, iaculis et justo et, imperdiet consectetur nisl. In magna 
nisl, aliquam vel faucibus ac, dapibus ut nisl. 

\subsection{Vectorization}
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a 
nibh pulvinar, ultricies sapien eget, consectetur massa. In augue 
ante, iaculis et justo et, imperdiet consectetur nisl. In magna 
nisl, aliquam vel faucibus ac, dapibus ut nisl. 

\subsection{Matrix implementation}
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a 
nibh pulvinar, ultricies sapien eget, consectetur massa. In augue 
ante, iaculis et justo et, imperdiet consectetur nisl. In magna 
nisl, aliquam vel faucibus ac, dapibus ut nisl. 

\subsection{Benchmark on BxD recombinant inbred lines}
Data of N phenotypes was extracted from GeneNetwork, together with 
X gene extression datasets, and the recently update genotypes for 
the BxD family. PCC within and between gene expression datasets are 
computed to build networks of co-expressed genes. PCC between 
classical phenotypes and gene expression is commonly used to 
indentify genes which might be involved in phenotype regulation. 
Correlation between genotypes is used to build genetic maps and 
build haplotypes. We benchmarked two different scenarios to show 
the improvement of MPCC over the R $cor()$ function.

{\bf Scenario 1:} Computation of PCC between all available genotypes 
in the BxD strain. Since genotype data in the BxD family is almost 
complete data, with only a few heterozygous (missing) loci remaining, 
this scenario benchmarks MPCC versus $cor()$ when a limited amount of 
missing data is present (Fig 1a).

{\bf Scenario 2:} Computation of PCC between classical phenotypes and 
multiple gene expression data sets available in GeneNetwork. This 
benchmark scenario compared MPCC versus $cor()$ in the presence of 
a lot of missing data since phenotypes are measured on a limited set 
of strains. Gene expression datasets are added in a step-wise fashion 
to increase the computational requirements in a step-wise fashion 
(Fig 1b).

\end{methods}

\section{Discussion}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna.

\section{Conclusion}

Driven by demands in large data biology we optimized correlations to a new level. 
The MPCC algorithm will greatly reduce the bottleneck for many fields aiming to 
use this application at large problem sizes. This is highly relevant for 
bioinformatics since so many tools use these statistics. A 100x speedup may sound
modest, however, it means getting diagnostics 100x faster, or use 1 computer 
instead of 100 significantly reducing costs. Our version can be run from any 
computer language, including R.

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero.

\section*{Acknowledgements}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam.
\vspace*{-12pt}

\section*{Funding}

This work has been supported by the... Text Text  Text Text.\vspace*{-12pt}

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}

%\bibliographystyle{plain}
\bibliography{main}

\end{document}
