\documentclass{bioinfo}
\copyrightyear{2019} \pubyear{2019}

\access{Advance Access Publication Date: Day Month 2019}
\appnotes{Application Note}

\begin{document}
\firstpage{1}

\subtitle{High-performance Computing}

\title[MPCC]{Matrix Pearson Correlation Coefficients}
\author[Arends \textit{et~al}.]{
Danny Arends\,$^{\text{\sfb 1, $\dagger$}}$, 
Mitch Horton\,$^{\text{\sfb 2, $\dagger$}}$ 
Chad Burdyshaw\,$^{\text{\sfb 2, $\dagger$}}$ 
Christian Fischer\,$^{\text{\sfb 3}}$ 
Pjotr Prins\,$^{\text{\sfb 3}}$ 
Rob W. Williams\,$^{\text{\sfb 3,*}}$ 
, and Glen Brook\,$^{\text{\sfb 2,*}}$}
\address{$^{\text{\sf 1}}$Z{\"u}chtungsbiologie und molekulare 
Genetik, Albrecht Daniel Thaer-Institut, Berlin, 10115, Germany \\
$^{\text{\sf 2}}$The Joint Institute for Computational Sciences, 
University of Tennessee, Oak 
Ridge, TN 37830, USA\\
$^{\text{\sf 3}}$Genetics, Genomics and Informatics, University 
of Tennessee Health Science Center, Memphis, TN 38163, USA.}

\corresp{$^\dagger$Contributed equally and should be considered 
joined first authors, $^\ast$To whom correspondence should be 
addressed.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{\textbf{Motivation:} Pearons correlation coefficients (PCC) 
is one of the most used correlation algorithms. However, computational 
efficiency of current implementations is low when computing many 
PCCs in the presence of missing data. This work was undertaken to 
improve the performance of PCC computation and allow the algorithm 
to be used on large biological data set currently available in the 
field of bioinformatics and computational biology.\\
\textbf{Results:} In this paper we present an C++ algorithm for fast 
computation of PCCs using matrix multiplication and vectorization. 
The implementation is independant of the amount of missing data 
present. Benchmarks againt the cor() function available in the R show 
that our package sigificantly reduces the runtime with up to 90 fold 
reduction in computational time on a single machine.\\
\textbf{Availability:} Code is available under an GPL-v3 
licence for C++ and The R Project for Statistical 
Computing at \href{https://github.com/UTennessee-JICS/MPCC}{https://github.com/UTennessee-JICS/MPCC}\\
\textbf{Contact:} \href{rwilliams@uthsc.edu}{rwilliams@uthsc.edu} or 
\href{glenn-brook@tennessee.edu}{glenn-brook@tennessee.edu}\\
\textbf{Supplementary information:} Supplementary data are 
available at \textit{Bioinformatics} online.}

\maketitle

\section{Introduction}
Pearons correlation coefficients introduced by Pearson in 19XX 
\citep{Bag01}, is one of the most used correlation algorithms 
in science. It is used ubiquitously in all field of science 
ranging from agriculture to zoology. Examples of large scale 
correlation matrices can be found in many areas of biology and 
bioinformatics. Genotype correlation are commonly used to 
construct haplotypes, build genetic maps, and order markers 
within the genome. Furthermore, PCCs have been used in (genome 
wide) association analysis, and novel algortihms to reconstruct 
genetic networks such as correlated trait locus (CTL) mapping. 
Many online databases such as 
GeneNetwork (\href{https://genenetwork.org/}{https://genenetwork.org/}) 
or the 
Mouse Gene Expression Database (GXD) (\href{https://informatics.jax.org/}{https://informatics.jax.org/})
provide large amounts of phenotype and gene expression data as 
well as online analysis tools to compute PCCs between e.g. 
gene expression within or between different tissues.\par

The cor() function available in R, is implemented in C / C++ and 
is relatively performant when no missing data is present. However, 
in the presence of missing data 5 different options on how-to deal 
with missing data are provided. The main choices are ($complete.obs$), 
which removes all rows that contain missing data, essentially 
creating a dataset without missing data. This is often unwanted in 
bioinformatics because missing data is often omni-present in data 
sets. The second main choice is the ($pairwise.complete.obs$) option, 
which due to it's implementation causes branching on missing data. This
leads to cache inefficiencies on the CPU, significanly increasing the 
runtime of the computation.
%\enlargethispage{12pt}

\section{Approach}
The correlaton algorithm was optimized, and binding to the commonly 
used R language were created. The R function is a 'drop in' replacement 
for the current cor() function.\par
Benchmarks show that computation of correlations between 10k genetic 
markers (genotypes) shows a 90 fold reduction in computational time 
compared to the standard cor function available in the R language, as 
such this new algorithm allows for larger datasets (30k x 30k) to be 
analyzed, as well as speed up analysis on medium scale datasets 
(1k x 1k).\par
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a 
nibh pulvinar, ultricies sapien eget, consectetur massa. In augue 
ante, iaculis et justo et, imperdiet consectetur nisl. In magna 
nisl, aliquam vel faucibus ac, dapibus ut nisl. Donec sollicitudin 
convallis vehicula. Fusce id est ut neque blandit convallis eu eget 
orci. Interdum et malesuada fames ac ante ipsum primis in faucibus. 

\begin{methods}
\section{Methods}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a 
nibh pulvinar, ultricies sapien eget, consectetur massa. In augue 
ante, iaculis et justo et, imperdiet consectetur nisl. In magna 
nisl, aliquam vel faucibus ac, dapibus ut nisl. Donec sollicitudin 
convallis vehicula. Fusce id est ut neque blandit convallis eu eget 
orci. Interdum et malesuada fames ac ante ipsum primis in faucibus. 

\begin{itemize}
\item for bulleted list, use itemize
\item for bulleted list, use itemize
\item for bulleted list, use itemize\vspace*{1pt}
\end{itemize}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\subsection{This is subheading}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\begin{table}[!t]
\processtable{This is table caption\label{Tab:01}} {\begin{tabular}{@{}llll@{}}\toprule head1 &
head2 & head3 & head4\\\midrule
row1 & row1 & row1 & row1\\
row2 & row2 & row2 & row2\\
row3 & row3 & row3 & row3\\
row4 & row4 & row4 & row4\\\botrule
\end{tabular}}{This is a footnote}
\end{table}

\end{methods}

\begin{figure}[!tpb]%figure1
\fboxsep=0pt\colorbox{gray}{\begin{minipage}[t]{235pt} \vbox to 100pt{\vfill\hbox to
235pt{\hfill\fontsize{24pt}{24pt}\selectfont FPO\hfill}\vfill}
\end{minipage}}
%\centerline{\includegraphics{fig01.eps}}
\caption{Caption, caption.}\label{fig:01}
\end{figure}

%\begin{figure}[!tpb]%figure2
%%\centerline{\includegraphics{fig02.eps}}
%\caption{Caption, caption.}\label{fig:02}
%\end{figure}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\subsection{This is subheading}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\section{Discussion}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\section{Conclusion}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis.

\begin{enumerate}
\item this is item, use enumerate
\item this is item, use enumerate
\item this is item, use enumerate
\end{enumerate}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 


Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 


\section*{Acknowledgements}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna.
\vspace*{-12pt}

\section*{Funding}

This work has been supported by the... Text Text  Text Text.\vspace*{-12pt}

%\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}


\begin{thebibliography}{}

\bibitem[Bofelli {\it et~al}., 2000]{Boffelli03}
Bofelli,F., Name2, Name3 (2003) Article title, {\it Journal Name}, {\bf 199}, 133-154.

\bibitem[Bag {\it et~al}., 2001]{Bag01}
Bag,M., Name2, Name3 (2001) Article title, {\it Journal Name}, {\bf 99}, 33-54.

\bibitem[Yoo \textit{et~al}., 2003]{Yoo03}
Yoo,M.S. \textit{et~al}. (2003) Oxidative stress regulated genes
in nigral dopaminergic neurnol cell: correlation with the known
pathology in Parkinson's disease. \textit{Brain Res. Mol. Brain
Res.}, \textbf{110}(Suppl. 1), 76--84.

\bibitem[Lehmann, 1986]{Leh86}
Lehmann,E.L. (1986) Chapter title. \textit{Book Title}. Vol.~1, 2nd edn. Springer-Verlag, New York.

\bibitem[Crenshaw and Jones, 2003]{Cre03}
Crenshaw, B.,III, and Jones, W.B.,Jr (2003) The future of clinical
cancer management: one tumor, one chip. \textit{Bioinformatics},
doi:10.1093/bioinformatics/btn000.

\bibitem[Auhtor \textit{et~al}. (2000)]{Aut00}
Auhtor,A.B. \textit{et~al}. (2000) Chapter title. In Smith, A.C.
(ed.), \textit{Book Title}, 2nd edn. Publisher, Location, Vol. 1, pp.
???--???.

\bibitem[Bardet, 1920]{Bar20}
Bardet, G. (1920) Sur un syndrome d'obesite infantile avec
polydactylie et retinite pigmentaire (contribution a l'etude des
formes cliniques de l'obesite hypophysaire). PhD Thesis, name of
institution, Paris, France.

\end{thebibliography}
\end{document}
