\documentclass{bioinfo}
\copyrightyear{2019} \pubyear{2019}

\access{Advance Access Publication Date: Day Month 2019}
\appnotes{Application Note}

\begin{document}
\firstpage{1}

\subtitle{Genetic and population analysis}

\title[MPCC]{Matrix Pearson Correlation Coefficients (MPCC) - An R Package for Performant Computation of Pearson Correlations with Missing Data Support}
\author[Arends \textit{et~al}.]{
Danny Arends\,$^{\text{\sfb 1, $\dagger$}}$, 
Mitch Horton\,$^{\text{\sfb 2, $\dagger$}}$ 
Chad Burdyshaw\,$^{\text{\sfb 2, $\dagger$}}$ 
Christian Fischer\,$^{\text{\sfb 3}}$ 
Pjotr Prins\,$^{\text{\sfb 3}}$ 
Rob W. Williams\,$^{\text{\sfb 3, *}}$ 
, and Glen Brook\,$^{\text{\sfb 2, *}}$}
\address{$^{\text{\sf 1}}$Z{\"u}chtungsbiologie und molekulare 
Genetik, Albrecht Daniel Thaer-Institut, Berlin, 10115, Germany \\
$^{\text{\sf 2}}$The Joint Institute for Computational Sciences, 
University of Tennessee, Oak 
Ridge, TN 37830, USA\\
$^{\text{\sf 3}}$Genetics, Genomics and Informatics, University 
of Tennessee Health Science Center, Memphis, TN 38163, USA.}

\corresp{$^\dagger$Contributed equally and should be considered 
joined first authors, $^\ast$To whom correspondence should be 
addressed.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{\textbf{Motivation:} Pearons correlation coefficients (PCC) 
is one of the most used correlation algorithms. However, computational 
efficiency of current implementations is low when computing many 
PCCs in the presence of missing data. This work was undertaken to 
improve the performance of PCC computation and allow the algorithm 
to be used on large biological data set currently available in the 
field of bioinformatics and computational biology.\\
\textbf{Results:} In this paper we present an C++ algorithm for fast 
computation of PCCs using matrix multiplication and vectorization. 
The implementation is independant of the amount of missing data 
present. Benchmarks againt the $cor()$ function available in the R show 
that our package sigificantly reduces the runtime with up to 90 fold 
reduction in computational time on a single machine.\\
\textbf{Availability:} Code is available under an GPL-v3 
licence for C++ and The R Project for Statistical 
Computing at \href{https://github.com/UTennessee-JICS/MPCC}{https://github.com/UTennessee-JICS/MPCC}\\
\textbf{Contact:} \href{rwilliams@uthsc.edu}{rwilliams@uthsc.edu} or 
\href{glenn-brook@tennessee.edu}{glenn-brook@tennessee.edu}\\
\textbf{Supplementary information:} Supplementary data are 
available at \textit{Bioinformatics} online.}

\maketitle

\section{Introduction}
Biology has driven innovation in statistics from the early days. Pearson, Fischer 
and Galton other famous statisticians were all working on biological data. The 
mathematical formula for Pearson correlation were derived by Auguste Bravais in 
1844. However, as Stigler's Law \citep{Stigler1980} dictates the methods name is 
crediting Karl Pearson, who was building on ideas published by Francis Galton in 
the 1880s. 

Pearson correlation is one of the most used correlation algorithms in science. 
It is used ubiquitously in all field of science ranging from agriculture to 
zoology. Examples of large scale correlation computation can be found in many 
areas of biology and bioinformatics. Genotype correlation are commonly used to 
construct haplotypes, build genetic maps, and order markers within the genome. 
Furthermore, PCCs have been used in co-expression analysis \citep{Tesson:2010}, 
(genome wide) association analysis, and novel algortihms to reconstruct genetic 
networks such as DiffCor \citep{Fukushima:2013}, weighted correlation network 
analysis (WGCNA) \citep{Horvath:2008} and correlated trait locus (CTL) mapping 
\citep{Arends2016a}.

Many online databases such as GeneNetwork 
(\href{https://genenetwork.org/}{https://genenetwork.org/}) \citep{Sloan2016} or 
the Mouse Gene Expression Database (GXD) 
(\href{https://informatics.jax.org/}{https://informatics.jax.org/}) \citep{Finger2017}
provide large amounts of genotype, classical phenotype, and gene expression data as 
well as online analysis tools such as PCCs. For these platforms, improvements in 
PCC computation allow for reduced costs, and improved waiting time for users.

PCCs computation in the R language for statistical computing \citep{R:2005} is 
provided by the $cor()$ function. The computation is implemented in C / C++ and 
is relatively performant when no missing data is present. However, 
in the presence of missing data 5 different options on how-to deal 
with missing data are provided. The main choices are ($complete.obs$), 
which removes all rows that contain missing data, essentially 
creating a dataset without missing data. This is often unwanted in 
bioinformatics because missing data is often omni-present in data 
sets. The second main choice is the ($pairwise.complete.obs$) option, 
which due to it's implementation causes branching on missing data. This
leads to cache inefficiencies on the CPU, significanly increasing the 
runtime of the computation.
%\enlargethispage{12pt}

\section{Approach}
We optimized the PCC algorithm by using matrix-matrix multiplication, 
optimized CPU cache efficiency, and using the MKL \citep{Intel2007, Wang2014} 
for optimal usage of intel CPU architecture, after which bindings to the R language 
were created. The new R function is a 'drop in' replacement for the 
$cor()$ function provided by R.

Benchmarking showed that computing correlations between 10k genetic markers 
(genotypes) shows a ~ 90 fold reduction in computational time compared to 
the standard R $cor()$ function. This mean the algorithm presented here
allows for larger datasets to be analyzed, as well as speed up analysis on 
small to medium scale datasets.

\begin{methods}
\section{Methods}

We started with the naive version of the algorithm, a series of vector-vector 
operations inside a double nested loop. We hand-tuned the naive version to the 
maximum extent attainable. Additionally, we reformulated the problem into a series of 
matrix-matrix operations. These operations call Intelâ€™s Math Kernel Library (MKL) 
directly. MKL comes, out of the box, tuned for high performance on emerging 
Intel architectures.

\begin{itemize}
\item Missing data masking
\item Memory access optimizations
\item Threading
\item Vectorization
\item Matrix implementation
\end{itemize}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a 
nibh pulvinar, ultricies sapien eget, consectetur massa. In augue 
ante, iaculis et justo et, imperdiet consectetur nisl. In magna 
nisl, aliquam vel faucibus ac, dapibus ut nisl. Donec sollicitudin 
convallis vehicula. Fusce id est ut neque blandit convallis eu eget 
orci. Interdum et malesuada fames ac ante ipsum primis in faucibus. 

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\subsection{This is subheading}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\begin{table}[!t]
\processtable{This is table caption\label{Tab:01}} {\begin{tabular}{@{}llll@{}}\toprule head1 &
head2 & head3 & head4\\\midrule
row1 & row1 & row1 & row1\\
row2 & row2 & row2 & row2\\
row3 & row3 & row3 & row3\\
row4 & row4 & row4 & row4\\\botrule
\end{tabular}}{This is a footnote}
\end{table}

\end{methods}

\begin{figure}[!tpb]%figure1
\fboxsep=0pt\colorbox{gray}{\begin{minipage}[t]{235pt} \vbox to 100pt{\vfill\hbox to
235pt{\hfill\fontsize{24pt}{24pt}\selectfont FPO\hfill}\vfill}
\end{minipage}}
%\centerline{\includegraphics{fig01.eps}}
\caption{Caption, caption.}\label{fig:01}
\end{figure}

%\begin{figure}[!tpb]%figure2
%%\centerline{\includegraphics{fig02.eps}}
%\caption{Caption, caption.}\label{fig:02}
%\end{figure}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\subsection{This is subheading}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\section{Discussion}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 

\section{Conclusion}

Driven by demands in large data biology we optimized correlations to a new level. 
The MPCC algorithm will greatly reduce the bottleneck for many fields aiming to 
use this application at large problem sizes. This is highly relevant for 
bioinformatics since so many tools use these statistics. A 100x speedup may sound
modest, however, it means getting diagnostics 100x faster, or use 1 computer 
instead of 100 significantly reducing costs. Our version can be run from any 
computer languge, including R.

\begin{enumerate}
\item this is item, use enumerate
\item this is item, use enumerate
\item this is item, use enumerate
\end{enumerate}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 


Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna. 
Aliquam ut diam sagittis, tempor mi eu, egestas libero. Duis sit 
amet dictum odio, a rutrum ex. Mauris eu cursus lacus. Donec 
lobortis non risus in blandit. Phasellus pellentesque molestie 
eros a mattis. 


\section*{Acknowledgements}

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent 
rhoncus ex vel enim volutpat, non volutpat est aliquam. Nullam est 
nunc, convallis congue efficitur ac, vestibulum sit amet magna.
\vspace*{-12pt}

\section*{Funding}

This work has been supported by the... Text Text  Text Text.\vspace*{-12pt}

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{main}

\end{document}
